### PR TITLE
GitCommands: Avoid creating a fake remote ref on pull

### PR DESCRIPTION
**Pull request by orgads**
This was done in eb3fce806cb76d328b669adf8de9a04244436172.

Passing +refs/heads/remote-branch:refs/remotes/origin/local-branch is
plain wrong. It creates a local ref for a remote branch which doesn't
really exist (origin/local-branch).

We don't really need a persistent ref for the remote branch. Just
passing it as a raw argument is fine. It is fetched into FETCH_HEAD,
then the local branch is merged/rebased on top of it.

### CODE REVIEW

**Comment by orgads** [comment]
ping

**Comment by vbjay** [comment]
There is no need to bump issues.

**Comment by orgads** [comment]
Then what should be done in order to get a PR reviewed? :)

**Comment by vbjay** [comment]
Be patient. https://github.com/gitextensions/gitextensions/pulse

**Comment by vbjay** [comment]
Also.  https://github.com/orgs/gitextensions/people are the only people
that can pull the commits into the repos.

**Comment by jbialobr** [comment]
>Then what should be done in order to get a PR reviewed? :)

You can find a PR that you would like to be merged and review it. Then you can ask the author of the PR you have reviewed to review your PR in exchange. I will merge reviewed and accepted PRs.

**Comment by orgads** [comment]
Rebased. Can anyone *PLEASE* review this?

**Comment by jbialobr** [comment]
I have done this. We need to add integration tests for GetFetchArgs #1948. To do this we need to refactor this method a bit.

Can you write a sketch of tests that fail for the case you are correcting? 

Something like:

var actual = GetFetchArgs(...)
string expected = "expectedResultHere";

**Comment by orgads** [comment]
Added unit tests.

File: UnitTests/GitCommandsTests/Git/GitCommandHelpersTest.cs [hunk 1]
```diff
@@ -59,6 +59,16 @@ public void CanGetRelativeNegativeDateString()
         }
 
         [TestMethod]
+        public void TestFetchArguments()
+        {
+            GitModule module = new GitModule(null);
+            var fetchCmd = module.FetchCmd("origin", "master", "local");
+            Assert.AreEqual("fetch --progress \"origin\" master --no-tags", fetchCmd);
```
Comment (jbialobr) [inline UnitTests/GitCommandsTests/Git/GitCommandHelpersTest.cs]: I believe you want this:
Assert.AreEqual("fetch --progress \"origin\" +refs/heads/master:local --no-tags", fetchCmd);
Do ypu want to fetch remote branch directly into local branch?

**Comment by orgads** [comment]
Strange. Here the test passes without refs/heads/ and fails with refs/heads/, but on appveyor it's the opposite. I'll investigate that on Sunday.

File: UnitTests/GitCommandsTests/Git/GitCommandHelpersTest.cs (hunk 1 above)
Comment (orgads) [reply UnitTests/GitCommandsTests/Git/GitCommandHelpersTest.cs]: No. The whole point of this change is _not_ to create a local branch on fetch! When you fetch master branch from origin, it updates the refs/remotes/origin/master. There's no need to create another ref for it.

I removed the localBranch argument which is not redundant.

**Comment by orgads** [comment]
I found out why it failed here. If the test is executed inside a git repo, and this repo has `origin` with `master` branch then it becomes `refs/heads/master`. Otherwise it remains `master`. I changed it to `some-branch` to avoid this inconsistency.

**Comment by orgads** [comment]
Ok, I discovered that the local branch is actually used on fetch with URL (I find it confusing, but whatever). Changed this patch to maintain this behavior, but never create a local branch on pull.

File: UnitTests/GitCommandsTests/Git/GitCommandHelpersTest.cs (hunk 1 above)
Comment (jbialobr) [reply UnitTests/GitCommandsTests/Git/GitCommandHelpersTest.cs]: Sorry, but I don't understand the idea of your change. If you want ignore the localBranch argument, so what is the point of filling it out on the Fetch dialog?

**Comment by jbialobr** [comment]
I accept the changes when it comes to the pull (merge, rebase) part. It seems that the changes made to `GetFetchArgs` are unnecessary to alter the pull command behavior. These changes cause inconsistency in FormPull dialog. User can fill the local branch name but it is never used in the git fetch command when fetching from a configured remote. I suggest to revert these changes as, if I am not mistaken, your original concern was about the pull command.

File: UnitTests/GitCommandsTests/Git/GitCommandHelpersTest.cs (hunk 1 above)
Comment (orgads) [reply UnitTests/GitCommandsTests/Git/GitCommandHelpersTest.cs]: Changed this in the last revision. Now I removed the argument only for pull, but left it for fetch.

**Comment by orgads** [comment]
[**@jbialobr**](https://github.com/jbialobr) commented on [Sep 20, 2017, 10:30 PM GMT+3](https://github.com/gitextensions/gitextensions/pull/3484#issuecomment-330956522 "2017-09-20T19:30:43Z - Replied by Github Reply Comments"):
> I accept the changes when it comes to the pull (merge, rebase) part. It seems that the changes made to `GetFetchArgs` are unnecessary to alter the pull command behavior. These changes cause inconsistency in FormPull dialog. User can fill the local branch name but it is never used in the git fetch command when fetching from a configured remote. I suggest to revert these changes as, if I am not mistaken, your original concern was about the pull command.

Please review the latest revision. I reverted the changes to GetFetchArgs, and now I only changed Pull's behavior.

**Comment by jbialobr** [comment]
I am talking about lines 1620-1623. I think that they should stay untouched, we should not remove them.

File: GitCommands/Git/GitModule.cs [hunk 2]
```diff
@@ -1612,15 +1612,11 @@ private string GetFetchArgs(string remote, string remoteBranch, string localBran
                 remoteBranchArguments = "+" + FormatBranchName(remoteBranch);
             }
 
-            string localBranchArguments;
+            string localBranchArguments = "";
             var remoteUrl = GetSetting(string.Format(SettingKeyString.RemoteUrl, remote));
 
             if (PathIsUrl(remote) && !string.IsNullOrEmpty(localBranch) && string.IsNullOrEmpty(remoteUrl))
                 localBranchArguments = ":" + GitCommandHelpers.GetFullBranchName(localBranch);
-            else if (string.IsNullOrEmpty(localBranch) || PathIsUrl(remote) || string.IsNullOrEmpty(remoteUrl))
```
Comment (jbialobr) [inline GitCommands/Git/GitModule.cs]: Passing null from PullCmd we will go to 1621 line and achieve the desired behavior. Discussion on how the fetch should act should be a subject of another PR in order to not block this PR.

Comment (orgads) [reply GitCommands/Git/GitModule.cs]: Done

**Comment by jbialobr** [comment]
Thank you.
